### ISSUE TITLE
Proposal: Introduce Feature Flags for Modular Build Configuration

### ISSUE THREAD

**Issue by joelonsql** [PostgREST/postgrest#3777]
Hello fellow PostgREST hackers,

I'm interested in exploring the possibility of introducing feature flags in PostgREST to allow users to exclude specific existing and future features at build time, similar to Rust's feature flags or C's `./configure --without-[feature]`. I believe this could make PostgREST more modular and customizable, enabling users to build it with only the functionalities they require. This approach could enhance security by reducing the attack surface and improve performance through lower memory consumption.

I'm curious if Haskell supports this kind of conditional compilation. I've read a bit about Cabal Flags but am unsure whether they would just disable code or actually exclude it from being built. I've also learned that Haskell can use the C Preprocessor (CPP) for conditional compilation. I'm not certain which approach would be best and would greatly appreciate any input or suggestions!

Personally, I'm particularly interested in building an RPC-only version of PostgREST. I've successfully built production systems that accessed PostgreSQL strictly via database functions, without needing the following functionalities:

- Resource-oriented REST API (Tables/Views)
- Query Parameter Operations
- Computed Fields
- Pagination and Count
- Resource Embedding
- Aggregate Functions
- JWT

I'm eager to contribute to this effort and would love to hear your thoughts on its feasibility and value to the project.

Thank you!

Joel Jacobson

**Comment by steve-chavez** [comment]
>  that accessed PostgreSQL strictly via database functions, without needing the following functionalities:
> That's a lot of stuff that you want to have disabled :)

Hm, yeah a lot to disable. @joelonsql Maybe we should start from what you actually need? Is it just to be able to call `/func` and specify arguments on the query string `/func?x=1&y=2`?

**Comment by steve-chavez** [comment]
This package was shared [before](https://github.com/PostgREST/postgrest/pull/872#issuecomment-304706193) on the subject of conditional compilation: https://hackage.haskell.org/package/plugins

**Comment by joelonsql** [comment]
> > that accessed PostgreSQL strictly via database functions, without needing the following functionalities:
> > That's a lot of stuff that you want to have disabled :)
> 
> Hm, yeah a lot to disable. @joelonsql Maybe we should start from what you actually need? Is it just to be able to call `/func` and specify arguments on the query string `/func?x=1&y=2`?

Yes, only need calling functions, passing arguments as a POST data in JSON.

**Comment by joelonsql** [comment]
> Could you elaborate on how you are securing access to this with a focus on security but without JWT support? Curious! :)

Using the `postgrest.conf` feature `pre-request = "public.auth"` and then good ol' Role-based access control (RBAC).
Here is a proof-of-concept: https://github.com/truthly/uniphant/blob/master/FUNCTIONS/auth.sql

I haven't worked on the uniphant repo for quite some time, so it might not be working, but should be enough to give an idea on the concept.

**Comment by joelonsql** [comment]
> I assume the intent here is to separate your data layer from the "access layer" or whatever you call it. So those database functions would be that layer. Note that, to have a nice HTTP api, it's **much** more convenient to actually use VIEWs for that layer as well and not limit yourself to RPCs. For RPCs you only have GET and POST, while for views, you have GET, POST, PATCH, DELETE - but you can still fully control all the logic in INSTEAD OF triggers. Those are kind of similar to your RPCs, but give you a cleaner interface.
> 
> We call this [Schema Isolation](https://docs.postgrest.org/en/v12/explanations/schema_isolation.html). I strongly suggest you think about that, too.

Yes! I forgot to mention, I actually use VIEWs also, but only for read operations, as I believe write operations are better handled via functions. When I need user-filtered data, I'm using the user_id GUC set by my RBAC implementation upon successful authentication and authorization, and then using that to filter VIEWs. Here is an example:

```sql
CREATE OR REPLACE VIEW api.current_user WITH (security_barrier) AS
SELECT
user_id,
username
FROM users
WHERE user_id = user_id();
```

I prefer this explicit filtering, rather than relying on RLS, which feels a bit too fragile and magical to me. I wish we would have had SELECT triggers instead of RLS, but that's another story.
